backend/src/test/java/com/google/minicrm/servlets/FormsServletTest.java

```diff
@@ -0,0 +1,283 @@
+  private static Map<String, Object> envAttributes;
+  static {
+    envAttributes = new HashMap<>();
+    envAttributes.put("com.google.appengine.api.users.UserService.user_id_key", TEST_USER_ID);
+  }
+  private static final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(
+          new LocalDatastoreServiceTestConfig(),
+          new LocalUserServiceTestConfig())
+          .setEnvIsLoggedIn(true)
+          .setEnvEmail("test@test.com")
+          .setEnvAuthDomain("testAuthDomain")
+          .setEnvAttributes(envAttributes);
+  private final Gson gson = new Gson();
+
+  private Form form1;
+  private Form form2;
+  private Form form3;
+  private HttpServletRequest request;
+  private HttpServletResponse response;
+
+  @Before
+  public void setUp() throws Exception {
+    helper.setUp();
+    request = mock(HttpServletRequest.class);
+    response = mock(HttpServletResponse.class);
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void formsServlet_getRequest_returnsAllFormsInTimeOrder() throws Exception {
+    slowSeedForms();
+    Form[] returnedForms = getForms();
+    Form[] expectedForms = {form3, form2, form1};
+    assertArrayEquals(expectedForms, returnedForms);
+  }
+
+  @Test
+  public void formsServlet_putRequest_json_successfullyRenamesAndReturns204() throws Exception {
+    seedForms();
+    when(request.getContentType()).thenReturn("application/json;");
+    Reader reader = new StringReader(new FormsPutRequest("2", "newName").toJson());
+    when(request.getReader()).thenReturn(new BufferedReader(reader));
+    formsServlet.doPut(request, response);
+
+    //get the new forms
+    List<Form> returnedForms = Arrays.asList(getForms());
+    form2.setFormName("newName");
+    List<Form> expectedForms = new ArrayList<>(3);
+    expectedForms.add(form1);
+    expectedForms.add(form2);
+    expectedForms.add(form3);
+    //assert all forms are the same regardless of order in list
+    assertTrue(expectedForms.size() == returnedForms.size());
+    assertTrue(expectedForms.containsAll(returnedForms) && expectedForms.containsAll(returnedForms));
+
+    //verify response codes
+    verify(response).setStatus(204);
+    verify(response, never()).sendError(Mockito.anyInt(), Mockito.anyString());
+    verify(response, never()).sendError(Mockito.anyInt());
+  }
+
+  @Test
+  public void formsServlet_putRequest_urlEncoded_successfullyRenamesAndReturns204()
+      throws Exception {
+    seedForms();
+    when(request.getContentType()).thenReturn("application/x-www-form-urlencoded;");
+    when(request.getParameter("formId")).thenReturn("2");
+    when(request.getParameter("formName")).thenReturn("newName");
+    formsServlet.doPut(request, response);
+
+    //get the new forms
+    List<Form> returnedForms = Arrays.asList(getForms());
+    form2.setFormName("newName");
+    List<Form> expectedForms = new ArrayList<>(3);
```

Comment:
This is a good idea, but initializing a List of a fixed size is probably not worth it performance-wise for such a small amount, and can make it harder to modify in the future.

Reply:
Will change.